addons/procurement/procurement.py

```diff
@@ -170,12 +170,9 @@ def onchange_product_id(self, cr, uid, ids, product_id, context=None):
             return {'value': v}
         return {}
 
-    def get_cancel_ids(self, cr, uid, ids, context=None):
-        return [proc.id for proc in self.browse(cr, uid, ids, context=context) if proc.state != 'done']
-
     def cancel(self, cr, uid, ids, context=None):
         #cancel only the procurements that aren't done already
-        to_cancel_ids = self.get_cancel_ids(cr, uid, ids, context=context)
+        to_cancel_ids = [proc.id for proc in self.browse(cr, uid, ids, context=context) if proc.state != 'done']
         if to_cancel_ids:
             return self.write(cr, uid, to_cancel_ids, {'state': 'cancel'}, context=context)
```

Comment:
When extract ids instead of just `to_cancel = self.browse(…).filtered(lambda p: p.state != 'done)` and calling `to_cancel.write({'state': 'cancel'})`?

Also this will blow up if it's called over RPC with no ids provided, is that intended?